File: app/src/main/java/cardgame/uxinn/is/cardgame/model/OverUnderGame.java
package cardgame.uxinn.is.cardgame.model;

import android.content.Context;
import android.util.Pair;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by aegir on 14.05.2018.
 */
public class OverUnderGame {
    private final Player player;
    private OverUnderDealer dealer;
    private List<OverUnderBet> bets;

    public OverUnderGame(Player player) {
        this.player = player;
    }

    public void startNewGame(Context context) {
        bets = new ArrayList<>();
        dealer = new OverUnderDealer();
        dealer.prepareGame();
    }

    public int remainingDeals() {
        return dealer.remainingDeals();
    }

    public Pair<Card, Card> deal() {
        if(remainingDeals() == 0) throw new RuntimeException("No more cards");
        return dealer.deal();
    }

    public void addBet(OverUnderBet bet) {
        bets.add(bet);
    }

    public List<OverUnderBet> getBets() {
        return bets;
    }

    public boolean hasMoreDeals() {
        return remainingDeals() > 0;
    }

    public boolean isPlayerWinner() {
        return false;
    }
}

File: app/src/main/java/cardgame/uxinn/is/cardgame/model/OverUnderGame.kt
package cardgame.uxinn.`is`.cardgame.model

import android.util.Pair
import java.util.*

/**
 * Created by aegir on 14.05.2018.
 */

data class Player(var wins: Int = 0)

class OverUnderGame(private val player: Player) {
    private lateinit var dealer: OverUnderDealer
    private lateinit var bets: MutableList<OverUnderBet>
    private var totalDeals: Int = 0

    val isFinished: Boolean
        get() = remainingDeals() == 0

    val isPlayerWinner: Boolean
        get() = player.wins > totalDeals / 2

    fun startNewGame() {
        bets = ArrayList()
        dealer = OverUnderDealer()
        dealer.prepareGame()
        totalDeals = dealer.remainingDeals()
        player.wins = 0
    }

    fun remainingDeals(): Int {
        return dealer.remainingDeals()
    }

    fun deal(): Pair<Card, Card> {
        if (remainingDeals() == 0) throw RuntimeException("No more cards")
        return dealer.deal()
    }

    fun addBet(bet: OverUnderBet) {
        bets.add(bet)
        if (bet.playerWins()) {
            player.wins++
        }
    }

    fun lastBet(): OverUnderBet {
        if (bets.size == 0) throw IndexOutOfBoundsException("No bets have been made")
        return bets[bets.size - 1]
    }
}
